### ISSUE TITLE
Dashboard not updating when browser window is not in focus.

### ISSUE THREAD

**Issue by Haukeeee** [webmin/webmin#1460]
Hi,

ever since the update to 1.973 with authentic theme 19.73 the server dashboard is not updating when I don't have the Firefox window in focus. It used to work and keeping a window open on second monitor with the dashboard was a nice way to see how the server was handling different tasks.

Is this intended behaviour or is something wrong on my end?

**Comment by iliarostovtsev** [comment]
Yes, it is expected. I made this change to prevent server over loads.

I was wondering if anyone would be interested in it. I could add an extra option, for **Enable real-time monitoring** called **Yes, always collect**.

**Comment by Haukeeee** [comment]
Thanks for the answer!
I would be very happy to have it as an option, maybe as opt-in so by default server load is reduced.

**Comment by iliarostovtsev** [comment]
Imagine what happens if you open 10 tabs with this option enabled. You would need to be aware about what's happening, when using this option.

**Comment by Haukeeee** [comment]
Could always have a disclaimer next to the option.
"Enabling real-time monitoring in the background can lead to a aheavy load on the host. Enable this only if you know what you are doing."

**Comment by iliarostovtsev** [comment]
It's always not easy to combine nice UI and usability. How about this:

![image](https://user-images.githubusercontent.com/4426533/111469897-4aee9480-8738-11eb-844a-22c0897f4614.png)

**Comment by iliarostovtsev** [comment]
Okay, latest commit adds it.

However it will require either restore theme to defaults or open browser's console and run:


```
settings_sysinfo_real_time_status = 2
theme.config.save()
```

**Comment by Haukeeee** [comment]
Thank you very much, much appreciated!

**Comment by SamSirry** [comment]
@iliaross, when the forced collection is done, I assume it is done just like the old behavior (before 1.973), right?

To be more specific, my worry is that the stats are collected even after all webmin pages are closed. I think this is not true, but doesn't harm to ask, I hope.

On the other hand, I can suggest a workaround for the server overload issue by caching the result on the server for a few seconds. Something like this:
- Webpage requests stats.
- Server looks for recently cached stats.
    - If stats are fresh enough they're sent to client.
    - If cached stats aren't found, or stale, the server collects stats and records them with a timestamp in a temporary cache file.

The only decision that needs to be done is how old is considered stale, and there's already such an option in the theme configuration that determines how often to pull stats.

Atomic operation is needed on the server for when a new stats request comes before the previous one was done processing...
What do you think?

**Comment by iliarostovtsev** [comment]
> when the forced collection is done, I assume it is done just like the old behavior (before 1.973), right?

No, there was a bug that restarted (re-enabled) stats collection even though it wasn't disabled on page focus, generating a long queue for nothing (creating overload).

> To be more specific, my worry is that the stats are collected even after all webmin pages are closed. I think this is not true, but doesn't harm to ask, I hope.

No, it is not possible, as it sends _SIGTERM_ on browser close.

> On the other hand, I can suggest a workaround for the server overload issue by caching the result on the server for a few seconds. Something like this:

Live stats are live, and suppose no caching. Status collection done by Webmin is cached, and refreshed by default every 5 minutes.

> Atomic operation is needed on the server for when a new stats request comes before the previous one was done processing

We currently don't use sockets, and client won't know about what's happening unless makes a call to the server.

> I can suggest a workaround for the server overload issue by caching the result on the server for a few seconds.

I really don't see any issues happening with overload in the future.

If you want to change query rate, you could do it (with theme 19.73+) by running the following command in the browser's console:

~~~
settings_sysinfo_query_timeout = 2000
theme.config.save()
~~~

**Comment by iliarostovtsev** [comment]
The right solution for this would be is to create a daemon to collect system info even when browser window is not focused.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
